### PR TITLE
RUM-3175 Fix possible file name conflict during consent change

### PR DESCRIPTION
**Pull request by ncreated**
### What and why?

📦 This PR addresses a potential file name conflict issue during tracking consent changes.

Initially considered a flakiness ❄️ in [testWhenWritingEventsWithPendingConsentThenGranted_itUploadsAllEvents()](https://github.com/DataDog/dd-sdk-ios/blob/004ca93c6036c0ceead2f09fedebfae9496c9da3/DatadogCore/Tests/Datadog/DatadogCore/DatadogCoreTests.swift#L81) (added in https://github.com/DataDog/dd-sdk-ios/pull/2063), it was discovered to be a real issue that can occur in certain edge cases.

The conflict happens when files are moved between `.pending` and `.granted` folders and events are written immediately before and after the consent change. If this occurs very quickly (under 1ms), the last batch in the pending folder may be named the same as the next batch in the granted folder. Since the SDK [does not check if the file exists](https://github.com/DataDog/dd-sdk-ios/blob/004ca93c6036c0ceead2f09fedebfae9496c9da3/DatadogCore/Sources/Core/Storage/Files/Directory.swift#L111-L115), the migrated file gets overwritten by the new event:

```
T - timestamp

[T0] consent: PENDING
[T1] write event A --> batch file created: intermediate/T1
[T1] change consent to GRANTED --> move intermediate/T1 to authorized/T1
[T1] write event B --> batch file created: authorized/T1 
                                           ^ 🐞 The previous T1 with event A is deleted
                                            and replaced with the new batch for event B.
```

### How?

This fix ensures that batch file names are unique, even in the edge case above, by waiting and generating a new file name when a conflict is detected. It guarantees that the new file's timestamp will be at least one precision interval (1ms) later than the existing file, preventing conflicts during consent changes.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [x] Add CHANGELOG entry for user facing changes
- [ ] Add Objective-C interface for public APIs (see our [guidelines](https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/3157787243/RFC+-+Modular+Objective-C+Interface#Recommended-solution) [internal]) and run `make api-surface`)

### CODE REVIEW

**Review by mariedm — COMMENTED** [review]
Great job identifying this issue! 💪

File: DatadogCore/Sources/Core/Storage/FilesOrchestrator.swift [hunk 1]
```diff
@@ -118,6 +123,26 @@ internal class FilesOrchestrator: FilesOrchestratorType {
...
+    /// Conflicts could only occur in very specific scenarios, such as during a tracking consent change when files are moved
+    /// from an unauthorized (.pending) folder to an authorized (.granted) folder, with events being written immediately before
+    /// and after the consent change. These conflicts were observed in tests, causing flakiness. In real-device scenarios,
+    /// conflicts can occur if tracking consent is changed and two events are written within the precision window defined
+    /// by `Constants.fileNamePrecission` (1 millisecond).
+    private func nextFileName() -> String {
+        var newFileName = fileNameFrom(fileCreationDate: dateProvider.now)
+        while directory.hasFile(named: newFileName) {
+            // Wait for the precision duration to avoid generating the same file name
+            Thread.sleep(forTimeInterval: Constants.fileNamePrecision)
```
[the remark this answers is not in the document]
Comment (maxep) [reply DatadogCore/Sources/Core/Storage/FilesOrchestrator.swift:140]: I'm also not a fan of adding a sleep here. Since the date is used to sort batches at upload, can't we just add a 1ms to the `dateProvider.now` instead?

[the remark this answers is not in the document]
Comment (ncreated) [reply DatadogCore/Sources/Core/Storage/FilesOrchestrator.swift]: Makes sense, I've updated the code to add `1ms` in case of a conflict 👍 ✅.

@mariedm, appending a random ID could have more indirect consequences here. On upload trigger, we read the recent batches and sort them by date. Introducing an ID would make the ordering ambiguous if two timestamps are the same, which would require additional code changes to handle these edge cases. For that reason, adjusting the timestamp seems like a less impactful solution.

**Review by mariedm — APPROVED** [review]

File: DatadogCore/Sources/Core/Storage/FilesOrchestrator.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (mariedm) [reply DatadogCore/Sources/Core/Storage/FilesOrchestrator.swift]: In that case I agree, using a random ID doesn't make sense.

**Review by maxep — APPROVED** [review]
Great catch!
